src/serac/physics/utilities/variational_form/weak_form.hpp

```diff
@@ -0,0 +1,472 @@
+ * my_residual.AddVolumeIntegral(integrand, domain_of_integration);
+ * // alternatively...
+ * my_residual.AddDomainIntegral(Dimension<3>{}, integrand, domain_of_integration);
+ * @endcode
+ */
+template <typename test, typename trial>
+class WeakForm<test(trial)> : public mfem::Operator {
+public:
+  /**
+   * @brief Constructs using @p mfem::ParFiniteElementSpace objects corresponding to the test/trial spaces
+   * @param[in] test_fes The test space
+   * @param[in] trial_fes The trial space
+   */
+  WeakForm(mfem::ParFiniteElementSpace* test_fes, mfem::ParFiniteElementSpace* trial_fes)
+      : Operator(test_fes->GetTrueVSize(), trial_fes->GetTrueVSize()),
+        test_space_(test_fes),
+        trial_space_(trial_fes),
+        P_test_(test_space_->GetProlongationMatrix()),
+        G_test_(test_space_->GetElementRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC)),
+        G_test_boundary_(
+            test_space_->GetFaceRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC, mfem::FaceType::Boundary)),
+        P_trial_(trial_space_->GetProlongationMatrix()),
+        G_trial_(trial_space_->GetElementRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC)),
+        G_trial_boundary_(
+            trial_space_->GetFaceRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC, mfem::FaceType::Boundary)),
+        grad_(*this)
+  {
+    SLIC_ERROR_IF(!G_test_, "Couldn't retrieve element restriction operator for test space");
+    SLIC_ERROR_IF(!G_trial_, "Couldn't retrieve element restriction operator for trial space");
+
+    input_L_.SetSize(P_trial_->Height(), mfem::Device::GetMemoryType());
+    input_E_.SetSize(G_trial_->Height(), mfem::Device::GetMemoryType());
+    input_E_boundary_.SetSize(G_trial_boundary_->Height(), mfem::Device::GetMemoryType());
+
+    output_E_.SetSize(G_test_->Height(), mfem::Device::GetMemoryType());
+    output_E_boundary_.SetSize(G_test_boundary_->Height(), mfem::Device::GetMemoryType());
+    output_L_.SetSize(P_test_->Height(), mfem::Device::GetMemoryType());
+    output_L_boundary_.SetSize(P_test_->Height(), mfem::Device::GetMemoryType());
+
+    my_output_T.SetSize(test_fes->GetTrueVSize(), mfem::Device::GetMemoryType());
+
+    dummy_.SetSize(trial_fes->GetTrueVSize(), mfem::Device::GetMemoryType());
+  }
+
+  /**
+   * @brief Adds an integral term to the weak formulation of the PDE
+   * @tparam geometry_dim The dimension of the element (2 for quad, 3 for hex, etc)
+   * @tparam spatial_dim The full dimension of the mesh
+   * @param[in] integrand The user-provided quadrature function, see @p Integral
+   * @param[in] domain The domain on which to evaluate the integral
+   * @note The @p Dimension parameters are used to assist in the deduction of the @a geometry_dim
+   * and @a spatial_dim template parameter
+   */
+  template <int geometry_dim, int spatial_dim, typename lambda>
+  void AddIntegral(Dimension<geometry_dim>, Dimension<spatial_dim>, lambda&& integrand, mfem::Mesh& domain)
```

Comment:
Dumb template question: how does having a templated type give us more than a simple int?

Reply:
Ooof! That is ugly! I agree the typed option is better.